crypto/fipsmodule/evp/digestsign.c

```diff
@@ -275,3 +275,22 @@ int EVP_DigestVerify(EVP_MD_CTX *ctx, const uint8_t *sig, size_t sig_len,
   }
   return ret;
 }
+
+void EVP_MD_CTX_set_pkey_ctx(EVP_MD_CTX *ctx, EVP_PKEY_CTX *pctx)
+{
+    // pctx could be null, so we have to deal with the cleanup job here.
+    if (!(ctx->flags & EVP_MD_CTX_FLAG_KEEP_PKEY_CTX)) {
```

Comment:
I wonder how you will ever reach this code?

Let's say you execute `EVP_MD_CTX_set_pkey_ctx` with `pctx != NULL`. Okay, we now reference this through `ctx->pctx` and the `EVP_MD_CTX_FLAG_KEEP_PKEY_CTX` bit is set.  

I consider `env_md_ctx_st` nominally opaque (in practice it is not, because the definition is in `digest.h`), so there is no way to unset the `EVP_MD_CTX_FLAG_KEEP_PKEY_CTX` bit without calling this function with `pctx = NULL`. But then you would have had to free the memory pointed to by `ctx->pct` a priori to the function execution. How to do this without access to struct flag??

I guess the point is that the consumer has to manage a pointer to `pctx` in addition to `ctx->pctx`. Loosing this pointer, would mean you leak memory.

Having a function that mutates `EVP_MD_CTX_FLAG_KEEP_PKEY_CTX` can also be hazardous... It would be easy to end up with dangling pointers... In fact, if the consumer maintains a second reference, `ctx->pctx` will be dangling if that reference is used to free the memory.

You should explain this behaviour in the function description, as well as who has the responsibility of this memory life-cycle.

Reply:
And some tests for these edge cases...

Reply:
Added a small description in the comments, let me know if I need to add anything more.